File: src/windows/Maple.WinUI/MapleUI/Atoms/MuiText.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;

namespace Maple.UI.Atoms
{
    /// <summary>The nine type-scale steps (text.md § Variants).</summary>
    public enum MuiTextVariant
    {
        SourceTitle, SheetTitle, RowLabel, Body, ToolLabel, ChipLabel, Eyebrow, ValueChip, Filename,
    }

    /// <summary>Text color role (text.md § Props: color).</summary>
    public enum MuiTextColorRole { Main, Muted, OnAccent, Success, Warning, Error }

    /// <summary>
    /// Maple.UI Text atom (docs/design/maple-ui/components/text.md) — the
    /// styled text block every label/caption/title/body copy routes through.
    /// Wraps an internal <see cref="TextBlock"/> (a ContentControl composite,
    /// the same shape MuiButton/MuiLink use) rather than subclassing
    /// TextBlock directly, so this atom's own <see cref="Text"/> property is
    /// free to react to every set (needed for the Eyebrow variant's uppercase
    /// transform) without depending on overriding a built-in DP's metadata.
    /// </summary>
    public sealed class MuiText : ContentControl
    {
        public static readonly DependencyProperty TextProperty =
            DependencyProperty.Register(nameof(Text), typeof(string), typeof(MuiText),
                new PropertyMetadata(string.Empty, (d, _) => ((MuiText)d).Rebuild()));

        public static readonly DependencyProperty VariantProperty =
            DependencyProperty.Register(nameof(Variant), typeof(MuiTextVariant), typeof(MuiText),
                new PropertyMetadata(MuiTextVariant.Body, (d, _) => ((MuiText)d).Rebuild()));

        public static readonly DependencyProperty ColorRoleProperty =
            DependencyProperty.Register(nameof(ColorRole), typeof(MuiTextColorRole), typeof(MuiText),
                new PropertyMetadata(MuiTextColorRole.Main, (d, _) => ((MuiText)d).Rebuild()));

        public static readonly DependencyProperty TruncateProperty =
            DependencyProperty.Register(nameof(Truncate), typeof(bool), typeof(MuiText),
                new PropertyMetadata(false, (d, _) => ((MuiText)d).Rebuild()));

        public static readonly DependencyProperty LineClampProperty =
            DependencyProperty.Register(nameof(LineClamp), typeof(int?), typeof(MuiText),
                new PropertyMetadata(null, (d, _) => ((MuiText)d).Rebuild()));

        public string Text
        {
            get => (string)GetValue(TextProperty);
            set => SetValue(TextProperty, value);
        }

        public MuiTextVariant Variant
        {
            get => (MuiTextVariant)GetValue(VariantProperty);
            set => SetValue(VariantProperty, value);
        }

        public MuiTextColorRole ColorRole
        {
            get => (MuiTextColorRole)GetValue(ColorRoleProperty);
            set => SetValue(ColorRoleProperty, value);
        }

        /// <summary>Single-line ellipsis truncation.</summary>
        public bool Truncate
        {
            get => (bool)GetValue(TruncateProperty);
            set => SetValue(TruncateProperty, value);
        }

        /// <summary>Multi-line clamp before an ellipsis; ignored when
        /// <see cref="Truncate"/> is also set (text.md § Props).</summary>
        public int? LineClamp
        {
            get => (int?)GetValue(LineClampProperty);
            set => SetValue(LineClampProperty, value);
        }

        private readonly TextBlock _inner = new();

        public MuiText()
        {
            Content = _inner;
            IsTabStop = false;
            Rebuild();
        }

        private static string StyleKey(MuiTextVariant variant) => variant switch
        {
            MuiTextVariant.SourceTitle => "MuiTextSourceTitleStyle",
            MuiTextVariant.SheetTitle => "MuiTextSheetTitleStyle",
            MuiTextVariant.RowLabel => "MuiTextRowLabelStyle",
            MuiTextVariant.ToolLabel => "MuiTextToolLabelStyle",
            MuiTextVariant.ChipLabel => "MuiTextChipLabelStyle",
            MuiTextVariant.Eyebrow => "MuiTextEyebrowStyle",
            MuiTextVariant.ValueChip => "MuiTextValueChipStyle",
            MuiTextVariant.Filename => "MuiTextFilenameStyle",
            _ => "MuiTextBodyStyle",
        };

        private static Brush R(string key) => (Brush)Application.Current.Resources[key];

        private static Brush ColorFor(MuiTextColorRole role) => role switch
        {
            MuiTextColorRole.Muted => R("MapleTextMuted"),
            // text.md: no dedicated "on accent" token exists yet — text_main
            // already reads correctly over the primary/error fills used
            // elsewhere (per button.md's primary-fill label color).
            MuiTextColorRole.OnAccent => R("MapleTextMain"),
            MuiTextColorRole.Success => R("MapleSuccessText"),
            MuiTextColorRole.Warning => R("MapleWarn"),
            MuiTextColorRole.Error => R("MapleErrorText"),
            _ => R("MapleTextMain"),
        };

        private void Rebuild()
        {
            _inner.Style = (Style)Application.Current.Resources[StyleKey(Variant)];
            _inner.Foreground = ColorFor(ColorRole);

            // eyebrow (text.md § Variants): "uppercase, wide tracking" — no
            // CSS text-transform equivalent on TextBlock, so the transform is
            // applied to the string itself.
            _inner.Text = Variant == MuiTextVariant.Eyebrow ? (Text ?? string.Empty).ToUpperInvariant() : Text ?? string.Empty;

            if (Truncate)
            {
                _inner.TextTrimming = TextTrimming.CharacterEllipsis;
                _inner.TextWrapping = TextWrapping.NoWrap;
                _inner.MaxLines = 1;
            }
            else if (LineClamp is { } clamp && clamp > 0)
            {
                _inner.TextTrimming = TextTrimming.CharacterEllipsis;
                _inner.TextWrapping = TextWrapping.Wrap;
                _inner.MaxLines = clamp;
            }
            else
            {
                _inner.TextTrimming = TextTrimming.None;
                _inner.TextWrapping = TextWrapping.Wrap;
                _inner.MaxLines = 0;
            }

            // text.md § Accessibility: truncated/clamped text must still be
            // reachable in full — surface it as a tooltip (the closest
            // equivalent this platform has to a `title` attribute) whenever
            // truncation is active, so the tail isn't silently lost.
            if (Truncate || LineClamp is > 0)
                ToolTipService.SetToolTip(_inner, Text);
            else
                ToolTipService.SetToolTip(_inner, null);
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/Atoms/MuiText.swift
// MuiText.swift — Maple UI Text atom.
// Contract: docs/design/maple-ui/components/text.md

import SwiftUI

/// The full type scale (text.md §Variants), one case per named scale step.
public enum MuiTextVariant: Sendable {
    case sourceTitle
    case sheetTitle
    case rowLabel
    case body
    case toolLabel
    case chipLabel
    case eyebrow
    case valueChip
    case filename
}

/// Color roles (text.md §Props).
public enum MuiTextColor: Sendable {
    case main, muted, onAccent, success, warning, error
}

/// A styled text block — the single atom every label, caption, title, and
/// body copy routes through (text.md §Purpose).
///
/// Carries no semantic heading level of its own (text.md §Accessibility) —
/// callers applying `sourceTitle` for an `<h1>`-equivalent are responsible
/// for the platform's own structural role (e.g. `.accessibilityAddTraits(.isHeader)`).
public struct MuiText: View {
    public let content: String
    public let variant: MuiTextVariant
    public let color: MuiTextColor
    public let truncate: Bool
    public let lineClamp: Int?
    public let block: Bool

    public init(
        _ content: String,
        variant: MuiTextVariant = .body,
        color: MuiTextColor = .main,
        truncate: Bool = false,
        lineClamp: Int? = nil,
        block: Bool = false
    ) {
        self.content = content
        self.variant = variant
        self.color = color
        self.truncate = truncate
        self.lineClamp = lineClamp
        self.block = block
    }

    public var body: some View {
        Text(content)
            .font(MuiTokens.TypeScale.font(variant))
            .tracking(variant == .eyebrow ? 0.6 : 0)
            .textCase(variant == .eyebrow ? .uppercase : nil)
            .foregroundStyle(foregroundColor)
            .lineLimit(effectiveLineLimit)
            .truncationMode(.tail)
            .frame(maxWidth: block ? .infinity : nil, alignment: .leading)
    }

    private var effectiveLineLimit: Int? {
        if truncate { return 1 }
        if let lineClamp { return lineClamp }
        return nil
    }

    private var foregroundColor: Color {
        switch color {
        case .main: return MuiTokens.textMain
        case .muted: return MuiTokens.textMuted
        // No dedicated "on accent" token exists yet — text_main already
        // reads correctly over the primary/error accent fills used
        // elsewhere (text.md §Tokens used, mirroring button.md's primary
        // fill label color).
        case .onAccent: return MuiTokens.textMain
        case .success: return MuiTokens.successText
        case .warning: return MuiTokens.warn
        case .error: return MuiTokens.errorText
        }
    }
}

#Preview("MuiText — Type scale") {
    VStack(alignment: .leading, spacing: 10) {
        MuiText("Source Title", variant: .sourceTitle)
        MuiText("Sheet Title", variant: .sheetTitle)
        MuiText("Row label", variant: .rowLabel)
        MuiText("Body copy runs here.", variant: .body)
        MuiText("Tool label", variant: .toolLabel)
        MuiText("Chip label", variant: .chipLabel)
        MuiText("Eyebrow", variant: .eyebrow)
        MuiText("42.0", variant: .valueChip)
        MuiText("IMG_0001.dng", variant: .filename)
    }
    .padding()
    .background(MuiTokens.bg)
}

#Preview("MuiText — Color roles") {
    VStack(alignment: .leading, spacing: 8) {
        MuiText("Main", color: .main)
        MuiText("Muted", color: .muted)
        MuiText("Success", color: .success)
        MuiText("Warning", color: .warning)
        MuiText("Error", color: .error)
    }
    .padding()
    .background(MuiTokens.bg)
}

#Preview("MuiText — Truncation") {
    VStack(alignment: .leading, spacing: 12) {
        MuiText("This is a very long single line of text that should truncate with an ellipsis", truncate: true, block: true)
            .frame(width: 200)
        MuiText("This is a very long run of text that should clamp after two lines instead of one, showing an ellipsis at the end.", lineClamp: 2, block: true)
            .frame(width: 200)
    }
    .padding()
    .background(MuiTokens.bg)
}
